llvm/include/llvm/Telemetry/Telemetry.h

```diff
@@ -0,0 +1,136 @@
+  // telemetry data (Could be stdout, stderr, or some local paths, etc).
+  //
+  // These strings will be interpreted by the vendor's code.
+  // So the users must pick the from their vendor's pre-defined
+  // set of Destinations.
+  std::vector<std::string> AdditionalDestinations;
+};
+
+/// For isa, dyn_cast, etc operations on TelemetryInfo.
+typedef unsigned KindType;
+/// This struct is used by TelemetryInfo to support isa<>, dyn_cast<>
+/// operations.
+/// It is defined as a struct (rather than an enum) because it is
+/// expectend to be extended by subclasses which may have
+/// additional TelemetryInfo types defined to describe different events.
+struct EntryKind {
+  static const KindType Base = 0;
+};
+
+/// TelemetryInfo is the data courier, used to move instrumented data
+/// from the tool being monitored to the Telemery framework.
+///
+/// This base class contains only the basic set of telemetry data.
+/// Downstream implementations can add more fields as needed to describe
+/// additional events.
+///
+/// For example, The LLDB debugger can define a DebugCommandInfo subclass
+/// which has additional fields about the debug-command being instrumented,
+/// such as `CommandArguments` or `CommandName`.
+struct TelemetryInfo {
+  // This represents a unique-id, conventionally corresponding to
+  // a tool's session - i.e., every time the tool starts until it exits.
+  //
+  // Note: a tool could have multiple sessions running at once, in which
+  // case, these shall be multiple sets of TelemetryInfo with multiple unique
+  // ids.
+  //
+  // Different usages can assign different types of IDs to this field.
+  std::string SessionId;
+
+  TelemetryInfo() = default;
+  virtual ~TelemetryInfo() = default;
+
+  virtual json::Object serializeToJson() const;
+
+  // For isa, dyn_cast, etc, operations.
+  virtual KindType getKind() const { return EntryKind::Base; }
+  static bool classof(const TelemetryInfo *T) {
+    if (T == nullptr)
+      return false;
+    return T->getKind() == EntryKind::Base;
+  }
+};
+
+/// This class presents a data sink to which the Telemetry framework
+/// sends data.
+///
+/// Its implementation is transparent to the framework.
+/// It is up to the vendor to decide which pieces of data to forward
+/// and where to forward them.
+class Destination {
+public:
+  virtual ~Destination() = default;
+  virtual Error emitEntry(const TelemetryInfo *Entry) = 0;
+  virtual llvm::StringLiteral name() const = 0;
+};
+
+/// This class is the main interaction point between any LLVM tool
+/// and this framework.
+/// It is responsible for collecting telemetry data from the tool being
+/// monitored and transmitting the data elsewhere.
+class Telemeter {
+public:
+  // Invoked upon tool startup
+  virtual void atStartup(llvm::StringRef ToolPath, TelemetryInfo *Entry) = 0;
+
+  // Invoked upon tool exit.
+  virtual void atExit(llvm::StringRef ToolPath, TelemetryInfo *Entry) = 0;
```

Comment:
I'll try to paraphase the 3 questions (I think) posed in this comment:

1) Why does `atStartup` and `atExit` have to take a TelemetryInfo?
Answer: The intention is that the caller would put data into this TelemetryInfo and give to the `telemetry::Manager` via these `atStartup` or `atExit`.  In otherwords, the `TelemetryInfo` (and its subclasses) acts as the "data courier" to move data from the tool (the caller) to the telemtry framework.

2) Why would they need a `Toolpath` argument?
Answer: No, in fact, it doesn't. This piece of data could be put into the TelemetryInfo object,

3) does `atStartup` need to be separate function vs part of ctor (presumably of the `telemetry::Manager` class)

It could go either way.  My preference for having a separate `atStartup` from the ctor is that it allows for flexibility on when the `telemetry::Manager` is created. But that doesn't mean down stream implementations can't change this one way or the other.

Reply:
Okay, thanks for the explanation. I'm happy with the 2 and 3 responses. For the 1 case, how might `atStart` and `atExit` actually differ internally, since all the data is stored in the `TelemetryInfo` input? Relatedly, how would you send events at other points in time?

Reply:
>  For the 1 case, how might `atStart` and `atExit` actually differ internally, since all the data is stored in the `TelemetryInfo` input? 

Do you mean, how would the implementation of `atStart` and `atExit` be different?
At the base level, perhaps not much. But different tools could have different things it wants to do at startup VS at exit.

> Relatedly, how would you send events at other points in time?

The subclass of `telemetry::Manager` can define additional  functions for that (eg., for LLDB's , we define `lldb::telemetry::LldbManager::logMainExecutableLoad(ModuleTelemetryInfo)` to be called when the main executable is loaded, or `::logDebugCommand (DebugCommandTelemetryInfo)` to be called when a command is executed.) 

These functions would be specific to each tool, hence not shared in the `llvm::telemetry::Manager`  interface.
Also, their arguments (`ModuleTelemetryInfo` and `DebugCommandTelemetryInfo`) would be subclasses of `llvm::telemetry::TelemetryInfo`, as these classes can have additional fields/attributes to describe addition concepts.